aioredis/connection.py

```diff
@@ -552,6 +553,105 @@ async def read_response(self) -> Union[EncodableT, List[EncodableT]]:
         return cast(Union[EncodableT, List[EncodableT]], response)
 
 
+@enum.unique
+class _State(enum.IntEnum):
+    not_connected = enum.auto()
+    connected = enum.auto()
+    error = enum.auto()
+
+
+class RedisProtocol(asyncio.Protocol):
+    """ """
+
+    def __init__(self):
+        self._state = _State.not_connected
+        self._resp_queue = deque()
+
+        self._transport: Optional[asyncio.Transport] = None
+
+        self._parser = hiredis.Reader()
+
+        self._responder_task = None
+        self._exc = None
+        self._conn_waiter = asyncio.get_event_loop().create_future()
+
+    def connection_made(self, transport: asyncio.Transport) -> None:
+        self._transport = transport
+        sock = transport.get_extra_info("socket")
+        if sock is not None:
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+        self._state = _State.connected
+        self._conn_waiter.set_result(None)
+
+    @property
+    def connected(self) -> bool:
+        return self._state == _State.connected
+
+    async def wait_connected(self) -> None:
+        await self._conn_waiter
+
+    def send_command(self, cmd):
+        fut = asyncio.get_event_loop().create_future()
+
+        if self._state == _State.connected:
+            # It's possible that connection was dropped and connection_callback was not
+            # called yet, to stop spamming errors, avoid writing to broken pipe
+            # Both _UnixWritePipeTransport and _SelectorSocketTransport that we
+            # expect to see here have this attribute
```

Comment:
What about _ProactorSocketTransport? I don't know Windows so don't really know anything about it.